ContactManager2/src/components/ContactForm.jsx

```diff
@@ -0,0 +1,48 @@
+import { InputTextBox } from "./InputTextBox";
+import { EMAIL, FIRST_NAME, LAST_NAME, contactFieldValidator } from "../config/contactFormConfig";
+import { useCallback, useState } from "react";
+import PropTypes from "prop-types"
+import getUniqueId from "../utils/newId"
+
+export const ContactForm = ({ addContact }) => {
+  const [formData, setFormData] = useState({});
+  const [errorMessages, setErrorMessages] = useState({});
+
+  const updateField = useCallback((id, value) => {
+    setFormData(data => ({
+      ...data, [id]: value
+    }))
+    contactFieldValidator[id](value, setErrorMessage)
+  }, [])
+
+  const setErrorMessage = useCallback((id, errorMessage) => {
+    setErrorMessages(messages => ({
+      ...messages, [id]: errorMessage
+    }))
+  }, [])
+
+  const submitAction = (event) => {
+    event.preventDefault();
+    const errorMsgs = contactFieldValidator.validateAllFields(formData);
+    setErrorMessages(errorMsgs);
+    
+    const areAllErrorFieldsNull = Object.keys(errorMsgs).every(key => errorMsgs[key] === null)
+    if (areAllErrorFieldsNull) {
+      formData.id = getUniqueId();
+      addContact(formData);
+      setFormData({})
+    }
+  }
+
+  return <form className="form">
+    <InputTextBox fieldId={FIRST_NAME} value={formData[FIRST_NAME]} errorMessage={errorMessages[FIRST_NAME]} updateValue={updateField} />
+    <InputTextBox fieldId={LAST_NAME} value={formData[LAST_NAME]} errorMessage={errorMessages[LAST_NAME]} updateValue={updateField} />
+    <InputTextBox fieldId={EMAIL} value={formData[EMAIL]} errorMessage={errorMessages[EMAIL]} updateValue={updateField} />
```

Comment:
can we use map here ? as they all have same structure.

Reply:
done